### PR TITLE
Fix Products block tests

### PR DESCRIPTION
**Pull request by gigitux**
This PR fixes the Products block tests. The issue was that the `insertBlock` function adds the block after the last block: in the case of the template, the last block is the `footer` template parts. The legacy blocks don't render any products when added to a template-parts.

For this reason, I created the `insertBlockViaInserter` function. When the block is added via inserter, this latter is added after the focused block.

I'm not sure why the tests didn't fail on the CI. There was only a warning about the timeout.

image

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Run `npm run env:start` or `npm run env:restart`
2. Run `npm run test:e2e-pw`.
3. Ensure that all the tests don't fail.
4. Ensure that the CI doesn't fail

* [x] Do not include in the Testing Notes

### CODE REVIEW

**Comment by imanish003** [comment]
Hi @gigitux 👋 

Out of curiosity, I checked why the `editor.insertBlock` function provided by `@wordpress/e2e-test-utils-playwright` doesn't support inserting a block at a specific location.

Here is the code from the Gutenberg repository [here](https://github.com/WordPress/gutenberg/blob/5693a62214b6c76d3dc5f3f69d8aad187748af79/packages/e2e-test-utils-playwright/src/editor/insert-block.ts#L12-L40)

```jsx
/**
 * Insert a block.
 *
 * @param this
 * @param blockRepresentation Inserted block representation.
 */
async function insertBlock(
	this: Editor,
	blockRepresentation: BlockRepresentation
) {
	await this.page.evaluate( ( _blockRepresentation ) => {
		function recursiveCreateBlock( {
			name,
			attributes = {},
			innerBlocks = [],
		}: BlockRepresentation ): Object {
			return window.wp.blocks.createBlock(
				name,
				attributes,
				innerBlocks.map( ( innerBlock ) =>
					recursiveCreateBlock( innerBlock )
				)
			);
		}
		const block = recursiveCreateBlock( _blockRepresentation );

		window.wp.data.dispatch( 'core/block-editor' ).insertBlock( block );
	}, blockRepresentation );
}
```

The line responsible for inserting the block is:

```jsx
window.wp.data.dispatch( 'core/block-editor' ).insertBlock( block );
```

Upon further investigation, I discovered that the `insertBlock` function actually supports an index parameter. [Here is the relevant code:](https://github.com/WordPress/gutenberg/blob/5693a62214b6c76d3dc5f3f69d8aad187748af79/packages/block-editor/src/store/actions.js#L543-L572)

```jsx
/**
 * Action that inserts a single block, optionally at a specific index respective a root block list.
 *
 * Only allowed blocks are inserted. The action may fail silently for blocks that are not allowed or if
 * a templateLock is active on the block list.
 *
 * @param {Object}   block           Block object to insert.
 * @param {?number}  index           Index at which block should be inserted.
 * @param {?string}  rootClientId    Optional root client ID of block list on which to insert.
 * @param {?boolean} updateSelection If true block selection will be updated. If false, block selection will not change. Defaults to true.
 * @param {?Object}  meta            Optional Meta values to be passed to the action object.
 *
 * @return {Object} Action object.
 */
export function insertBlock(
	block,
	index,
	rootClientId,
	updateSelection,
	meta
) {
	return insertBlocks(
		[ block ],
		index,
		rootClientId,
		updateSelection,
		0,
		meta
	);
}
```

Here are a couple of suggestions:

1. We can contribute upstream to add support for the index parameter. This way, we won't need to create a new function to insert the block at a specific location.
2. In the meantime, we can create a copy of [this code](https://github.com/WordPress/gutenberg/blob/5693a62214b6c76d3dc5f3f69d8aad187748af79/packages/e2e-test-utils-playwright/src/editor/insert-block.ts#L12-L40) and make the necessary changes to support the index parameter. This will allow us to proceed without waiting for the changes to be merged upstream. Once the PR is merged upstream, we can remove the copied code.

What are your thoughts on this? 🙂

**Comment by gigitux** [comment]
> Hi @gigitux 👋
> 
> Out of curiosity, I checked why the `editor.insertBlock` function provided by `@wordpress/e2e-test-utils-playwright` doesn't support inserting a block at a specific location.
> 
> Here is the code from the Gutenberg repository [here](https://github.com/WordPress/gutenberg/blob/5693a62214b6c76d3dc5f3f69d8aad187748af79/packages/e2e-test-utils-playwright/src/editor/insert-block.ts#L12-L40)
> 
> ```js
> /**
>  * Insert a block.
>  *
>  * @param this
>  * @param blockRepresentation Inserted block representation.
>  */
> async function insertBlock(
> 	this: Editor,
> 	blockRepresentation: BlockRepresentation
> ) {
> 	await this.page.evaluate( ( _blockRepresentation ) => {
> 		function recursiveCreateBlock( {
> 			name,
> 			attributes = {},
> 			innerBlocks = [],
> 		}: BlockRepresentation ): Object {
> 			return window.wp.blocks.createBlock(
> 				name,
> 				attributes,
> 				innerBlocks.map( ( innerBlock ) =>
> 					recursiveCreateBlock( innerBlock )
> 				)
> 			);
> 		}
> 		const block = recursiveCreateBlock( _blockRepresentation );
> 
> 		window.wp.data.dispatch( 'core/block-editor' ).insertBlock( block );
> 	}, blockRepresentation );
> }
> ```
> 
> The line responsible for inserting the block is:
> 
> ```js
> window.wp.data.dispatch( 'core/block-editor' ).insertBlock( block );
> ```
> 
> Upon further investigation, I discovered that the `insertBlock` function actually supports an index parameter. [Here is the relevant code:](https://github.com/WordPress/gutenberg/blob/5693a62214b6c76d3dc5f3f69d8aad187748af79/packages/block-editor/src/store/actions.js#L543-L572)
> 
> ```js
> /**
>  * Action that inserts a single block, optionally at a specific index respective a root block list.
>  *
>  * Only allowed blocks are inserted. The action may fail silently for blocks that are not allowed or if
>  * a templateLock is active on the block list.
>  *
>  * @param {Object}   block           Block object to insert.
>  * @param {?number}  index           Index at which block should be inserted.
>  * @param {?string}  rootClientId    Optional root client ID of block list on which to insert.
>  * @param {?boolean} updateSelection If true block selection will be updated. If false, block selection will not change. Defaults to true.
>  * @param {?Object}  meta            Optional Meta values to be passed to the action object.
>  *
>  * @return {Object} Action object.
>  */
> export function insertBlock(
> 	block,
> 	index,
> 	rootClientId,
> 	updateSelection,
> 	meta
> ) {
> 	return insertBlocks(
> 		[ block ],
> 		index,
> 		rootClientId,
> 		updateSelection,
> 		0,
> 		meta
> 	);
> }
> ```
> 
> Here are a couple of suggestions:
> 
> 1. We can contribute upstream to add support for the index parameter. This way, we won't need to create a new function to insert the block at a specific location.
> 2. In the meantime, we can create a copy of [this code](https://github.com/WordPress/gutenberg/blob/5693a62214b6c76d3dc5f3f69d8aad187748af79/packages/e2e-test-utils-playwright/src/editor/insert-block.ts#L12-L40) and make the necessary changes to support the index parameter. This will allow us to proceed without waiting for the changes to be merged upstream. Once the PR is merged upstream, we can remove the copied code.
> 
> What are your thoughts on this? 🙂

Thanks, @imanish003, for your idea! It is great! I implemented your feedback and added two `@todo` comments:
1. The selector that I used works only with WordPress without Gutenberg enabled. I didn't have time to think about a scalable approach. My idea is to work on this when I will work on https://github.com/woocommerce/woocommerce/issues/42412
2. I added a `todo` comment to work on the Gutenberg PR.

File: tests/e2e-pw/utils/constants.ts
Comment (imanish003) [inline tests/e2e-pw/utils/constants.ts]: It appears that the TIMEOUT constants added in this PR are not being utilized. If that is indeed the case, perhaps we can consider removing these constants for now until they are required in the future. What do you think? 🙂

**Comment by imanish003** [comment]
It looks like the test is still failing but showing as PASSING: https://github.com/woocommerce/woocommerce-blocks/actions/runs/5378006279/jobs/9757234661?pr=9951

**Comment by gigitux** [comment]
> It looks like the test is still failing but showing as PASSING: https://github.com/woocommerce/woocommerce-blocks/actions/runs/5378006279/jobs/9757234661?pr=9951

It was flaky, but playwright re-try it and passed.

File: tests/e2e-pw/utils/constants.ts
Comment (gigitux) [reply tests/e2e-pw/utils/constants.ts]: Fixed via [d6c5284](https://github.com/woocommerce/woocommerce-blocks/pull/9951/commits/d6c52848b2380e2bac2346318f638b68e77ec8a2)

**Review by imanish003 — APPROVED** [review]
Hi @gigitux, Thanks for improving the flaky tests 🙌🏻

I ran the tests locally & all are passing 🚀

**Pull request merged**
